File: Website_Deploy/App_Code/abstract/pages/CPageWithTableHelpers.vb
Imports Microsoft.VisualBasic

Public Class CPageWithTableHelpers : Inherits CPage

    Protected Function Row(tbl As Table) As TableRow
        Dim tr As New TableRow
        tbl.Rows.Add(tr)
        If tbl.Rows.Count Mod 2 = 0 Then
            tr.CssClass = "alt_row"
        End If
        Return tr
    End Function
    Protected Function RowH(tbl As Table) As TableHeaderRow
        Dim tr As New TableHeaderRow
        tbl.Rows.Add(tr)
        Return tr
    End Function
    Protected Function RowH(tbl As Table, cellText As String, colspan As Integer) As TableHeaderRow
        Dim tr As TableHeaderRow = RowH(tbl)
        CellH(tr, cellText).ColumnSpan = colspan
        Return tr
    End Function
    Protected Function CellLinkH(row As TableRow, s As String, url As String, Optional black As Boolean = True, Optional bold As Boolean = True, Optional ttip As String = Nothing, Optional confirmMsg As String = Nothing) As TableHeaderCell
        Dim td As New TableHeaderCell
        row.Cells.Add(td)
        CellLink(td, s, url, black, bold, ttip, confirmMsg)
        Return td
    End Function
    Protected Function CellLinkR(row As TableRow, s As String, url As String, Optional black As Boolean = True, Optional bold As Boolean = True, Optional ttip As String = Nothing, Optional confirmMsg As String = Nothing) As TableCell
        Dim td As TableCell = CellLink(row, s, url, black, bold, ttip, confirmMsg)
        td.HorizontalAlign = HorizontalAlign.Right
        Return td
    End Function
    Protected Function CellLink(row As TableRow, s As String, url As String, Optional black As Boolean = True, Optional bold As Boolean = True, Optional ttip As String = Nothing, Optional confirmMsg As String = Nothing) As TableCell
        Dim td As New TableCell
        row.Cells.Add(td)
        Dim lnk As HyperLink = CellLink(td, s, url, black, bold, ttip, confirmMsg)
        Return td
    End Function
    Private Function CellLink(td As TableCell, s As String, url As String, black As Boolean, bold As Boolean, Optional ttip As String = Nothing, Optional confirmMsg As String = Nothing) As HyperLink
        If String.IsNullOrEmpty(s) Then Return New HyperLink

        Dim lnk As New HyperLink
        lnk.Text = s
        lnk.NavigateUrl = url
        td.Controls.Add(lnk)

        If Not String.IsNullOrEmpty(ttip) Then lnk.ToolTip = ttip
        If black Then lnk.ForeColor = Drawing.Color.Black
        If bold Then lnk.Font.Bold = True
        If Not IsNothing(confirmMsg) Then lnk.Attributes.Add("onclick", "return confirm('" & confirmMsg.Replace("'", "\'") & "')")
        Return lnk
    End Function
    Protected Function CellR(row As TableRow, Optional text As String = Nothing, Optional ttip As String = Nothing, Optional bold As Boolean = False) As TableCell
        CellR = Cell(row, text, ttip, bold)
        CellR.HorizontalAlign = HorizontalAlign.Right
    End Function
    Protected Function Cell(row As TableRow, Optional text As String = Nothing, Optional ttip As String = Nothing, Optional bold As Boolean = False) As TableCell
        Dim td As New TableCell
        row.Cells.Add(td)
        If Not String.IsNullOrEmpty(text) Then td.Text = text
        If Not String.IsNullOrEmpty(ttip) Then td.ToolTip = ttip
        If bold Then td.Font.Bold = True
        Return td
    End Function
    Protected Function CellH(row As TableRow, Optional text As String = Nothing, Optional ttip As String = Nothing, Optional bold As Boolean = False) As TableHeaderCell
        Dim th As New TableHeaderCell
        row.Cells.Add(th)
        If Not String.IsNullOrEmpty(text) Then th.Text = text
        If Not String.IsNullOrEmpty(ttip) Then th.ToolTip = ttip
        If bold Then th.Font.Bold = True
        Return th
    End Function
End Class

File: WebsiteCSharp/App_Code/abstract/pages/CPageWithTableHelpers.cs
using Microsoft.VisualBasic;
using System.Web.UI.WebControls;

public class CPageWithTableHelpers : CPage
{

    protected TableRow Row(Table tbl)
    {
        TableRow tr = new TableRow();
        tbl.Rows.Add(tr);
        if (((tbl.Rows.Count % 2)
                    == 0))
        {
            tr.CssClass = "alt_row";
        }

        return tr;
    }

    protected TableHeaderRow RowH(Table tbl)
    {
        TableHeaderRow tr = new TableHeaderRow();
        tbl.Rows.Add(tr);
        return tr;
    }

    protected TableHeaderRow RowH(Table tbl, string cellText, int colspan)
    {
        TableHeaderRow tr = this.RowH(tbl);
        this.CellH(tr, cellText).ColumnSpan = colspan;
        return tr;
    }

    protected TableHeaderCell CellLinkH(TableRow row, string s, string url, bool black = true, bool bold = true, string ttip = null, string confirmMsg = null)
    {
        TableHeaderCell td = new TableHeaderCell();
        // Warning!!! Optional parameters not supported
        // Warning!!! Optional parameters not supported
        // Warning!!! Optional parameters not supported
        // Warning!!! Optional parameters not supported
        row.Cells.Add(td);
        this.CellLink(td, s, url, black, bold, ttip, confirmMsg);
        return td;
    }

    protected TableCell CellLinkR(TableRow row, string s, string url, bool black = true, bool bold = true, string ttip = null, string confirmMsg = null)
    {
        TableCell td = this.CellLink(row, s, url, black, bold, ttip, confirmMsg);
        // Warning!!! Optional parameters not supported
        // Warning!!! Optional parameters not supported
        // Warning!!! Optional parameters not supported
        // Warning!!! Optional parameters not supported
        td.HorizontalAlign = HorizontalAlign.Right;
        return td;
    }

    protected TableCell CellLink(TableRow row, string s, string url, bool black = true, bool bold = true, string ttip = null, string confirmMsg = null)
    {
        TableCell td = new TableCell();
        // Warning!!! Optional parameters not supported
        // Warning!!! Optional parameters not supported
        // Warning!!! Optional parameters not supported
        // Warning!!! Optional parameters not supported
        row.Cells.Add(td);
        HyperLink lnk = this.CellLink(td, s, url, black, bold, ttip, confirmMsg);
        return td;
    }

    private HyperLink CellLink(TableCell td, string s, string url, bool black, bool bold, string ttip = null, string confirmMsg = null)
    {
        if (string.IsNullOrEmpty(s))
        {
            return new HyperLink();
        }

        // Warning!!! Optional parameters not supported
        // Warning!!! Optional parameters not supported
        HyperLink lnk = new HyperLink();
        lnk.Text = s;
        lnk.NavigateUrl = url;
        td.Controls.Add(lnk);
        if (!string.IsNullOrEmpty(ttip))
        {
            lnk.ToolTip = ttip;
        }

        if (black)
        {
            lnk.ForeColor = System.Drawing.Color.Black;
        }

        if (bold)
        {
            lnk.Font.Bold = true;
        }

        if (!(confirmMsg == null))
        {
            lnk.Attributes.Add("onclick", ("return confirm(\'"
                            + (confirmMsg.Replace("\'", "\\\'") + "\')")));
        }

        return lnk;
    }

    protected TableCell CellR(TableRow row,  string text= null, string ttip = null, bool bold = false)
    {
        var td = this.Cell(row, text, ttip, bold);
        td.HorizontalAlign = HorizontalAlign.Right;
        return td;
    }

    protected TableCell Cell(TableRow row, string text= null, string ttip = null, bool bold = false)
    {
        TableCell td = new TableCell();
        row.Cells.Add(td);
        if (!string.IsNullOrEmpty(text))
        {
            td.Text = text;
        }

        if (!string.IsNullOrEmpty(ttip))
        {
            td.ToolTip = ttip;
        }

        if (bold)
        {
            td.Font.Bold = true;
        }

        return td;
    }

    protected TableHeaderCell CellH(TableRow row, string text= null, string ttip = null, bool bold = false)
    {
        TableHeaderCell th = new TableHeaderCell();
        // Warning!!! Optional parameters not supported
        // Warning!!! Optional parameters not supported
        // Warning!!! Optional parameters not supported
        row.Cells.Add(th);
        if (!string.IsNullOrEmpty(text))
        {
            th.Text = text;
        }

        if (!string.IsNullOrEmpty(ttip))
        {
            th.ToolTip = ttip;
        }

        if (bold)
        {
            th.Font.Bold = true;
        }

        return th;
    }
}